docs/architecture/grpc-for-wcf-developers/apm.md

```diff
@@ -1,11 +1,278 @@
+
+```csharp
+_logger.LogError(exception, "Error opening file '{file}'.", filePath);
+```
+
+Using this syntax, the logger will format the message the same, but the `filePath` variable will be added as a named value `file` on a structured object passed along with the logs. Logging back-ends that support structured logging, such as [Serilog](https://serilog.net/) or [NLog](https://nlog-project.org/) (as of version 4.5) will write this structured object along with the message, as long as the storage target supports it.
+
+Despite having names, the tokens in the message format string are treated as positional parameters, just like in `string.Format`. The first token is treated as `{0}`, the second as `{1}` and so on. If a token is used more than once, the index of the first occurrence is used.
+
+### Storing log data
+
+Log entries are handled by *Providers*, which write the log messages out to various targets. ASP.NET Core logging provides some simple providers via NuGet packages, all prefixed with `Microsoft.Extensions.Logging`.
+
+| Provider | Description |
+| -------- | ----------- |
+| Console  | Writes simple messages to standard output |
+| Debug    | Writes simple messages to a debugger (e.g. Visual Studio) if any is attached |
+| TraceSource | Makes messages available to .NET TraceListener |
+| EventSource | Makes messages available to .NET EventListener |
+| EventLog | Writes messages to the Windows Event Log |
+| AzureAppServices | Writes messages and structured data to Azure "Diagnostic logs" and "Log stream" features |
+
+Writing to other storage systems, like SQL databases, text files, or specialized databases like [Elasticsearch](https://elastic.co), is not natively supported by the core libraries. Support for other platforms can be added using one of the various open source logging libraries available for .NET Core, such as Serilog, NLog or log4net, which have extensions supporting a wide range of logging targets. Refer to the documentation for your preferred library for information on how to configure it to work with ASP.NET Core logging.
+
+## Metrics in ASP.NET Core gRPC
+
+Unlike logging, there is no in-the-box support for producing metrics data in ASP.NET Core. There is an excellent open source library called [App Metrics](https://www.app-metrics.io), which provides an extensive set of types to instrument your code, and additional packages to write metrics to a variety of targets including time-series databases (e.g. Prometheus and InfluxDB), [Azure AppInsights](https://docs.microsoft.com/azure/azure-monitor/app/app-insights-overview) and more. The [App.Metrics.AspNetCore.Mvc](https://www.nuget.org/packages/App.Metrics.AspNetCore.Mvc/) package even adds a comprehensive set of basic metrics automatically generated via integration with the ASP.NET Core framework, and the web site provides templates for displaying those the [Grafana](https://grafana.com/) visualization platform.
```

Comment:
Is this still true? See https://devblogs.microsoft.com/dotnet/introducing-diagnostics-improvements-in-net-core-3-0/

gRPC uses the new counters to captures metrics on the [server](https://github.com/grpc/grpc-dotnet/blob/aad959b7292b0b98e99996bb056ffe73e494e635/src/Grpc.AspNetCore.Server/Internal/GrpcEventSource.cs) and [client](https://github.com/grpc/grpc-dotnet/blob/aad959b7292b0b98e99996bb056ffe73e494e635/src/Grpc.Net.Client/Internal/GrpcEventSource.cs).

It would still be useful to document how to listen to those counters and export them to a third-party.

// @shirhatti

Reply:
Added a note on EventSource/EventCounters.